src/mongo/db/storage/csv_file.cpp

```diff
@@ -0,0 +1,351 @@
+}
+
+template <>
+void appendTo<CsvFieldType::kDate>(BSONObjBuilder& builder,
+                                   const std::string& fieldName,
+                                   const std::string& data) {
+    auto date = dateFromISOString(data);
+    uassert(date.getStatus().code(), date.getStatus().toString(), date.isOK());
+    builder.appendDate(fieldName, date.getValue());
+}
+
+template <>
+void appendTo<CsvFieldType::kOid>(BSONObjBuilder& builder,
+                                  const std::string& fieldName,
+                                  const std::string& data) {
+    constexpr size_t kLengthOidValue = 24;
+    constexpr size_t kOidTypeStr = 8;                      // objectid
+    constexpr size_t kOidTypeStrPrefix = kOidTypeStr + 2;  // objectid("
+
+    std::string mutableData = data;
+    if (data[0] == 'O' || data[0] == 'o') {
+        std::transform(
+            mutableData.begin(), mutableData.begin() + kOidTypeStr, mutableData.begin(), ::tolower);
+        uassert(ErrorCodes::BadValue,
+                "Invalid Object Id format: {}"_format(data),
+                mutableData.substr(0, kOidTypeStrPrefix) == "objectid(\"");
+        mutableData = mutableData.substr(kOidTypeStrPrefix, kLengthOidValue);
+    } else if (data[0] == '\"') {
+        mutableData = mutableData.substr(1, kLengthOidValue);
+    }
+
+    uassert(ErrorCodes::BadValue,
+            "Invalid Object Id Format: {}"_format(data),
+            mutableData.length() == kLengthOidValue);
+    mongo::OID id = mongo::OID(mutableData);  // throws exception if it detects bad OID
+    builder.append(fieldName, id);
+}
+
+template <>
+void appendTo<CsvFieldType::kBool>(BSONObjBuilder& builder,
+                                   const std::string& fieldName,
+                                   const std::string& data) {
+    std::string mutableData = data;
+    bool val;
+    std::transform(data.begin(), data.end(), mutableData.begin(), ::tolower);
+    if (mutableData == "true" || mutableData == "t" || mutableData == "yes" || mutableData == "y" ||
+        mutableData == "1")
+        val = true;
+    else if (mutableData == "false" || mutableData == "f" || mutableData == "no" ||
+             mutableData == "n" || mutableData == "0")
+        val = false;
+    else
+        uasserted(ErrorCodes::BadValue, "{} is an invalid Boolean representation"_format(data));
+    builder.append(fieldName, val);
+}
+
+
+boost::optional<BSONObj> CsvFileInput::readBsonObj() {
+    std::string record;
+    std::getline(_ifs, record);
+    // if eof is reached, _ifs return false
+    if (_ifs.eof() || _ifs.fail()) {
+        return boost::none;
+    }
+
+    auto data = parseLine(record);
+    uassert(ErrorCodes::TypeMismatch,
+            "Metadata specifies {} fields but current record has {}"_format(_metadata.size(),
+                                                                            data.size()),
+            data.size() == _metadata.size());
+
```

Comment:
To be more permissive on user data, I think it would be better to allows cases where actual data has more fields than the metadata.

```suggestion
            data.size() >= _metadata.size());
```

Reply:
```suggestion
            if (data.size() != metadata.size()) {
                LOGV2_WARNING(200000407, "Metadata specifies {n1} fields but current record has {n2}", "n1"_attr = metadata.size, "n2"_attr = data.size())
            }

            n = min(data.size(), metadata.size());
            for (size_t i = 0; i < n; ++i) {.......}
```

Reply:
sgtm!